bpf: add feature flag for policy_map metrics

Jack-R-lantern opened the pull request:
Please ensure your pull request adheres to the following guidelines:

- [ ] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [ ] All code is covered by unit and/or runtime tests where feasible.
- [ ] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [ ] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [ ] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [ ] Provide a title or release-note blurb suitable for the release notes.
- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [ ] Thanks for contributing!

<!-- Description of change -->

Fixes: #28487 

```release-note
Add flag --policy-accounting to enable/disable per-policy packet and byte accounting (default true)
```

joestringer reviewed (changes requested):
Hi! Thanks for the PR.

Relaying some of the comments I provided on Slack: We can't just change the LPM type for the policymap here to make the statistics per-CPU. The LPM property provides a basis of some of the security decisionmaking in the policy engine, so this proposal will break some existing features.

If you're looking to make these statistics per-cpu, then there might be options to extend the statistics fields to be per-cpu and ensure each CPU accesses only its own field, or perhaps split stats out. But I would suggest by starting with just wrapping the stats behind a (new) feature flag.

Jack-R-lantern committed:
bpf: add POLICY_ACCOUNTING macro

Policy ACCOUNTING has a lot of overhead.
Need to make it possible to turn it on/off through settings.

Signed-off-by: Jack-R-lantern <tjdfkr2421@gmail.com>

LeiW000 commented on bpf/ep_config.h:
```diff
@@ -54,6 +54,7 @@ DEFINE_U32(POLICY_VERDICT_LOG_FILTER, 0xffff);
 #define CUSTOM_CALLS_MAP test_cilium_calls_custom_65535
 #define LOCAL_DELIVERY_METRICS
 #define CONNTRACK_ACCOUNTING
+#define POLICYTRACING_ACCOUNTING
```
Why do we define it by default? I think what we want is to not run __sync_fetch_and_add() by default.

joestringer replied:
For what it's worth, this file is only used for compile testing and not used in production. However I think there is a solid debugability reason to keep this tracking stats on policymap by default and only optimize for performance for users who find such performance details to be critical to their application.

Jack-R-lantern committed:
pkg/option: add specPolicyAccounting

add specPolicyAccounting

Signed-off-by: Jack-R-lantern <tjdfkr2421@gmail.com>

LeiW000 commented on pkg/option/runtime_options.go:
```diff
@@ -60,6 +61,11 @@ var (
 		Description: "Enable trace notifications",
 	}
 
+	specPolicyTracingAccounting = Option{
```
I guess that it should be "specPolicyAccounting" rather than "specPolicyTracingAccounting".

LeiW000 replied:
By the way, you put "rebase"  as the commit message. It seems that there is something wrong in your workspace causing it.

Jack-R-lantern replied:
@LeiW000 thanks review i'll fixed it, I'm struggling with the part of the current option that writes to the header file because I don't quite understand it. Can you help me?

LeiW000 replied:
Yes, it seems that you forgot to map specPolicyAccounting to PolicyAccounting. I just created a patch. You can cherry-pick it to try. 
https://github.com/LeiW000/cilium/commits/policy_accounting
193796b0613b4db0e8bc1da6f5b5c3419d37f84d

LeiW000 replied:
@Jack-R-lantern , could you give me write permission for your branch[ A-Performance-issue-with-calling-__sync_fetch_and_add](https://github.com/Jack-R-lantern/cilium/tree/ISSUE-28487/A-Performance-issue-with-calling-__sync_fetch_and_add)? Then, I can push my commit into this PR directly. 

By the way, the options are written into /run/cilium/state/[endpoint id]/ ep_config.h, which will be used to compile the bpf program.

Jack-R-lantern replied:
@LeiW000 thank you, Sure I invite you to

Commit:
Map specPolicyAccounting to PolicyAccounting

Signed-off-by: Leiw <steven.l.wang@linux.intel.com>

LeiW000 reviewed (approved):
It looks good to me now.

joestringer reviewed (approved):
The change LGTM, thanks for the submission!

joestringer commented on bpf/lib/policy.h:
```diff
@@ -15,9 +15,11 @@ static __always_inline int
 __account_and_check(struct __ctx_buff *ctx, struct policy_entry *policy,
```
I see the report from @LeiW000 (thanks for testing!), please mark ctx as `__maybe_unused`.
```suggestion
__account_and_check(struct __ctx_buff *ctx __maybe_unused, struct policy_entry *policy,
```

Jack-R-lantern replied:
@LeiW000 @joestringer thanks, I've made changes based on those reviews.

Jack-R-lantern committed:
bpf: policy.h __account_and_check fix

add ctx parameter __maybe_unused extension

Signed-off-by: Jack-R-lantern <tjdfkr2421@gmail.com>

LeiW000 reviewed (approved):
LGTM.

derailed reviewed (commented):
@Jack-R-lantern LG Thank you for this pr!
Do we need to update docs with this new config option?

Jack-R-lantern committed:
docs: add daemon flag

add policy-accounting option

Signed-off-by: Jack-R-lantern <tjdfkr2421@gmail.com>

joestringer reviewed (approved):
LGTM.

derailed reviewed (approved):
@Jack-R-lantern LG Thank you for the updates!

LeiW000 reviewed (approved):
LGTM.
